wkcuber/downsampling.py

```diff
@@ -30,6 +31,27 @@ def calculate_virtual_scale_for_target_mag(
     return cast(Tuple[float, float, float], tuple(scale_array))
 
 
+def calculate_default_max_mag_for_dataset(dataset_path: str, layer_name: str) -> Mag:
+    ds = WKDataset(dataset_path)
+    ds_size = ds.properties.data_layers[layer_name].get_bounding_box_size()
+    cube_length = (
+        ds.properties.data_layers[layer_name].wkw_magnifications[0].cube_length
+    )
+    return calculate_default_max_mag(ds_size, cube_length)
+
+
+def calculate_default_max_mag(
+    dataset_size: Tuple[int, int, int], cube_length: int
+) -> Mag:
+    cubes_per_dim = -(-np.array(dataset_size) // np.array(cube_length))
+    magnification = [
+        int(math.pow(2, int(math.log(s, 2)))) for s in cubes_per_dim
+    ]  # highest power of 2 smaller than cubes_per_dim
+    return Mag(
+        [min(max(m, 16), 512) for m in magnification]
+    )  # at least 16 and at most 512
```

Comment:
I have no good intuition on how large the typical datasets are and what magnifications are necessary. @philippotto is the order of magnitude correct?

Reply:
[just for the record, as we already talked about most of this in person]

I gave this a bit of thought and I think what is ultimately needed is a very small, lightweight representation of the entire dataset. For this, we can set an arbitrary threshold such as 100*100*100 vx. That would be 1 MB uncompressed (in total) and one slice of it would be 10 kB (ideal for a thumbnail). Making the dataset even smaller probably doesn't help much for most usecases.

I think, I would use such a logic as a general heuristic instead of considering the cube counts. To give an example for a "typical dataset", a rather big one could be around 100000 × 300000 × 8000. Going for 100 voxel per edge would result in a target mag 1024, 4096, 128.
However, I would add the following constraints:
- X and Y should be equivalent (that's what we see 99% for existing datasets). Therefore, I'd simply consider `max(100000, 300000)` for x and y (--> 4096, 4096, 128).
- Additionally, I'd follow the anisotropic downsampling approach we usually do (see the code for that logic). Therefore, the method could simply output `Mag(4096)` as the max mag and combined with https://github.com/scalableminds/webknossos-cuber/issues/157, the z dimension would be downsampled according to the scale.